Add `qk_dag_compose` to the C API

raynelfss opened the pull request:
<!--
⚠️  If you do not respect this template, your pull request will be closed.
⚠️  Your pull request title should be short detailed and understandable for all.
⚠️  Also, please add a release note file using reno if the change needs to be documented in the release notes.
⚠️  If your pull request fixes an open issue, please link to the issue. Use "Fixes #XXXX" if this PR *fully* closes the issue XXXX.  
☢️  If you used an AI tool to code this PR, add "AI tool used: <Name and version of the tool>". For example, "AI tool used: Microsoft Copilot Chat with GPT-5". Failing to disclose the use of AI tools may result in the PR being closed without further review.  


- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary
The following commits add ``qk_dag_compose`` to the C API, as a counterpart to the ``DAGCircuit::compose`` method.


### Details and comments
Closes #15196.

raynelfss committed:
Initial: Add `qk_dag_compose` to the C API

raynelfss committed:
Test: Add one test
The test is based on the ``test_compose_inorder`` python test, but it is incomplete since we cannot check the order of the operations just yet. However, we can easily check the number of operation nodes as a temporary way of checking the validity of the composition.

aaryav-3 commented on crates/cext/src/dag.rs:
```diff
@@ -706,6 +708,96 @@ pub unsafe extern "C" fn qk_dag_op_node_kind(dag: *const DAGCircuit, node: u32)
+
+    // Since we don't yet supposed vars in C, we can skip the include_captures check.
+    dag.compose(other_dag, qubits.as_deref(), clbits.as_deref(), false)
+        .expect("Error during circuit composition.");
```
quick clarification here, do we need to use the .expect() method here? is it valid considering the rust panic, on failure, may result in undefined behavior across the FFI boundary? I couldn't be sure as I don't know if our C API has a way to handle such panics. 
I also wanted to ask why it's otherwise not preferable to handle such cases by unpacking the PyResult type return of :meth:`DAGCircuit.compose()` and mapping to the newly defined exit codes?

raynelfss replied:
Your suggestion here is a really good idea, mapping to the newly defined error codes **should** be the way to go. However, for this case we would have to do a couple of things before we can get there.
1. We would need `DAGCircuit` to use Rust native errors, the same way that we're doing for CircuitData in #15279. You may ask why couldn't we just map with the current errors? The problem here is that these errors are aliases for Python exceptions, and in order to get information about them, such as the error message or the type, we would have to go through python first (even if the message originated from Rust). So that means, unless we have our C API currently connected to python, we would panic regardless.
2.  Why not just return the generic exit code that we made for `DAGCircuit` (`ExitCode::DAGError`)? This would work similarly to the current approach except that it is recoverable. However, we get much less information than we do with the current approach since at least we know the error happened during the run of the pass. 

**We could, however, implement a generic ``ExitCode::DAGComposeError`` that may also be salvageable and communicates just as much.** This would depend on whether we want this to be the case here, I would be open to it since this method modifies the ``QkDag`` directly.

aaryav-3 commented on crates/cext/src/dag.rs:
```diff
@@ -706,6 +708,96 @@ pub unsafe extern "C" fn qk_dag_op_node_kind(dag: *const DAGCircuit, node: u32)
+/// # Safety
+///
+/// Behavior is undefined if ``dag`` or ``other`` are not valid, non-null pointers to a ``QkDag``.
+/// If ``qubit`` nor ``clbit`` are NULL, it must contains a less or equal amount
```
small typo in the line, should be 
```rust
/// If neither ``qubit`` nor ``clbit`` are NULL, it must contain a less or equal amount
```

also, the underlying compose method mandates the use of only an EQUAL amount of wires in both the other circuit and the dag circuit as is in this line in :meth:`DAGCircuit.compose`:

```rust
        let qubit_map = match qubits {
            None => identity_qubit_map.clone(),
            Some(qubits) => {
                if qubits.len() != other.qubits.len() {
                    return Err(DAGCircuitError::new_err(concat!(
                        "Number of items in qubits parameter does not",
                        " match number of qubits in the circuit."
                    )));
                }
```

but here we are allowing <= values in code, and not raising the mismatch exit code on the less than condition, am I missing something here?

raynelfss replied:
Thank you for identifying the typo. As for the second half of the comment, the method ``DAGCircuit::compose`` allows for usage of less or equal amount of qubits for the ``other`` dag, compared to the main DAG. You can see that the first check in the method does just as much:
```rs
        if other.qubits.len() > self.qubits.len() || other.clbits.len() > self.clbits.len() {
            return Err(DAGCircuitError::new_err(
                "Trying to compose with another DAGCircuit which has more 'in' edges.",
            ));
        }
```
Because what this method basically does is to append this dag onto the main dag it should be impossible to do so with more input nodes than the main dag, but a smaller or equally sized `DAGCircuit` instance should be fine.

Now your confusion here lies with how we compare the ``qubits`` and ``clbits`` argument. Since these represent the order of the wires in terms of the other dag to explain how we attach them to the main dag, these need to explicitly be **of the same length as ``other``** when provided (Which is what the code snippet that you referenced explains). So this entry point here does not break that convention and if it did the pass itself would scream at us for not sticking to it.

I hope this explanation clears up any confusion you may have and thank you for reviewing :)

aaryav-3 commented on test/c/test_dag.c:
```diff
@@ -312,6 +312,97 @@ static int test_dag_endpoint_node_value(void) {
     return result;
 }
 
+static int test_dag_compose(void) {
```
Do you think here we can also have test cases to test whether the correct exit codes are raised on failure, the new ones added in specific?

raynelfss replied:
Yes, that is a great idea. I will add a test that does that.

eliarbel replied:
It would also be good to cover the `qubit` and `clbit` mapping code in `qk_dag_compose`.

raynelfss replied:
Added tests with permutation in c662160174ee980bdb302d39d7765646f9a9c143 and 0088092df31ebc84a14e55d47672404fc3ecc19e

aaryav-3 reviewed (commented):
Thank you so much for your contribution! This is a great, clean FFI implementation for the :meth:`DAGCircuit::compose` method. I had just some quick clarifications before this PR looks perfect to merge.

eliarbel commented on test/c/test_dag.c:
```diff
@@ -312,6 +312,97 @@ static int test_dag_endpoint_node_value(void) {
+    //              │  ┌───┐
+    //  rqr_1: |0>──┼──┤ X ├
+    //              │  ├───┤
+    //  rqr_2: |0>──┼──┤ Y ├  =
```
```suggestion
    //  rqr_2: |0>──┼──┤ Y ├
```

eliarbel commented on test/c/test_dag.c:
```diff
@@ -312,6 +312,97 @@ static int test_dag_endpoint_node_value(void) {
+    // lcr_1: 0 ═══════════════════════
+    QkExitCode res = qk_dag_compose(dag_left, dag_right, NULL, 0, NULL, 0);
+    if (res != QkExitCode_Success) {
+        result = RuntimeError;
```
Should we use this instead?

```suggestion
        result = EqualityError;
```

eliarbel commented on test/c/test_dag.c:
```diff
@@ -312,6 +312,97 @@ static int test_dag_endpoint_node_value(void) {
+
+    if (left_op_nodes + right_op_nodes != new_op_nodes) {
+        result = EqualityError;
+        printf("The operations did not get composed onto the left dag correcly. Expected %zu "
```
```suggestion
        printf("The operations did not get composed onto the left dag correctly. Expected %zu "
```

eliarbel commented on crates/cext/src/exit_codes.rs:
```diff
@@ -60,6 +60,12 @@ pub enum ExitCode {
     TranspilerError = 400,
+    /// DAG related error,
+    DAGError = 500,
+    /// The DAGs have mismatching qubit amounts during compose.
```
```suggestion
    /// The DAGs have mismatching qubit/clbit amounts during compose.
```

eliarbel commented on crates/cext/src/dag.rs:
```diff
@@ -706,6 +708,96 @@ pub unsafe extern "C" fn qk_dag_op_node_kind(dag: *const DAGCircuit, node: u32)
+/// @param num_clbits The number of qubits in ``clbits``.
+///
+/// @return an exit code.
+///
```
It would be nice to have an example.

raynelfss replied:
Added in bcb0bacb6713e9e2a27f712e1a252a10458bb56b

eliarbel commented on crates/cext/src/dag.rs:
```diff
@@ -706,6 +708,96 @@ pub unsafe extern "C" fn qk_dag_op_node_kind(dag: *const DAGCircuit, node: u32)
 }
 
+/// @ingroup QkDag
+/// Compose the ``other`` circuit onto the ``dag`` instance with the option of a subset of
```
```suggestion
/// Compose the ``other`` DAG onto the ``dag`` instance with the option of a subset
```

eliarbel commented on crates/cext/src/dag.rs:
```diff
@@ -706,6 +708,96 @@ pub unsafe extern "C" fn qk_dag_op_node_kind(dag: *const DAGCircuit, node: u32)
 
+/// @ingroup QkDag
+/// Compose the ``other`` circuit onto the ``dag`` instance with the option of a subset of
+/// of input wires of ``other`` being mapped onto a subset of output wires of this circuit.
```
```suggestion
/// of input wires of ``other`` being mapped onto a subset of output wires of ``dag``.
```

eliarbel commented on crates/cext/src/dag.rs:
```diff
@@ -706,6 +708,96 @@ pub unsafe extern "C" fn qk_dag_op_node_kind(dag: *const DAGCircuit, node: u32)
+    qubits: *const u32,
+    num_qubits: usize,
+    clbits: *const u32,
+    num_clbits: usize,
```
I would use `u32` to keep it consistent with the rest of the API functions that expose similar parameters (e.g. `qk_quantum_register_new`):

```suggestion
    num_qubits: u32,
    clbits: *const u32,
    num_clbits: u32,
```

eliarbel commented on crates/cext/src/dag.rs:
```diff
@@ -706,6 +708,96 @@ pub unsafe extern "C" fn qk_dag_op_node_kind(dag: *const DAGCircuit, node: u32)
+
+    let qubits: Option<Vec<ShareableQubit>> = if check_ptr(qubits).is_ok() {
+        let qubits: Box<[u32]> =
+            unsafe { Box::from_raw(slice_from_raw_parts(qubits, num_qubits).cast_mut()) };
```
This is probably a UB since the pointer wasn't created with `Box::into_raw`. Also, we don't want to take ownership, do we? 
I would go with something like: 
```suggestion
        let qubits = unsafe { std::slice::from_raw_parts(qubits, num_qubits as usize) };
```

eliarbel commented on crates/cext/src/dag.rs:
```diff
@@ -706,6 +708,96 @@ pub unsafe extern "C" fn qk_dag_op_node_kind(dag: *const DAGCircuit, node: u32)
+                let Some(qubit) = dag.qubits().get(Qubit(bit)) else {
+                    return Err(ExitCode::DAGComposeMissingBit);
+                };
+                Ok(qubit.clone())
```
Just a style suggestion (assuming `qubits` is a borrowed slice per the comment above):

```suggestion
        let new_qubits: Result<Vec<ShareableQubit>, ExitCode> = qubits
            .iter()
            .map(|bit| -> Result<ShareableQubit, ExitCode> {
                dag.qubits().get(Qubit(*bit)).cloned().ok_or(ExitCode::DAGComposeMissingBit)
```

eliarbel commented on crates/cext/src/dag.rs:
```diff
@@ -706,6 +708,96 @@ pub unsafe extern "C" fn qk_dag_op_node_kind(dag: *const DAGCircuit, node: u32)
+    };
+
+    let clbits: Option<Vec<ShareableClbit>> = if check_ptr(clbits).is_ok() {
+        let clbits: Box<[u32]> =
+            unsafe { Box::from_raw(slice_from_raw_parts(clbits, num_clbits).cast_mut()) };
+        let new_clbits: Result<Vec<ShareableClbit>, ExitCode> = clbits
+            .into_iter()
+            .map(|bit| -> Result<ShareableClbit, ExitCode> {
+                let Some(clbit) = dag.clbits().get(Clbit(bit)) else {
+                    return Err(ExitCode::DAGComposeMissingBit);
+                };
+                Ok(clbit.clone())
+            })
```
Please see the two comments about `qubits`, which apply also to this block.

raynelfss replied:
Fixed in a028b51f86907954c50549d559c75f3597af4802

eliarbel commented on crates/cext/src/dag.rs:
```diff
@@ -706,6 +708,96 @@ pub unsafe extern "C" fn qk_dag_op_node_kind(dag: *const DAGCircuit, node: u32)
+        None
+    };
+
+    // Since we don't yet supposed vars in C, we can skip the include_captures check.
```
```suggestion
    // Since we don't yet support vars in C, we can skip the inline_captures check.
```

eliarbel reviewed (commented):
Thanks Ray for doing this. I've left some inline comments. 

In addition, we are missing a release note for the new function.

raynelfss committed:
Apply suggestions from code review

Co-authored-by: Eli Arbel <46826214+eliarbel@users.noreply.github.com>

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into c-dag-compose

raynelfss committed:
Fix: Use `std::slice::from_raw_parts`.

eliarbel commented on crates/cext/src/dag.rs:
```diff
@@ -1243,6 +1244,105 @@ pub unsafe extern "C" fn qk_dag_get_instruction(
+    num_qubits: u32,
+    clbits: *const u32,
+    num_clbits: u32,
+) -> ExitCode {
```
#15320 added a parameter to specify variables copy mode, although we don't have classical variables support yet in the C API: 
https://github.com/Qiskit/qiskit/blob/e33feba98b869a0af80aab37b90bbb6a6296b101/crates/cext/src/dag.rs#L1488

Given this #14332, I'm wondering whether we want to reserve a bool parameter for `inline_captures` in this PR already, to promote API stability. @jakelishman ?

Cryoris replied:
As discussed offline we won't include this now.

raynelfss committed:
Fix: Add extra test with qubit re-ordering.

raynelfss committed:
Fix: Finalize test by adding clbit reordering.

raynelfss committed:
Fix: memory leak in test_dag.c

raynelfss committed:
Docs: Add example in docstring

raynelfss committed:
Fix: Incorrect comparison
- Remove classical registers when not needed.

raynelfss committed:
Fix: C format

Cryoris commented on crates/cext/src/dag.rs:
```diff
@@ -1243,6 +1244,153 @@ pub unsafe extern "C" fn qk_dag_get_instruction(
+/// @param other A pointer to the DAG to compose with ``dag``.
+/// @param qubits A list of indices representing the qubit wires to compose
+///     onto.
+/// @param num_qubits The number of qubits in ``qubits``.
```
Don't we know the number of qubits and clbits from the `other` DAG already and can remove these arguments? We should then only add as safety comment that the `qubits` and `clbits` pointers should be readable for the appropriate number of elements.

raynelfss replied:
That's true. I will add a patch to fix it.

raynelfss replied:
Removed in a1306ec5cb4eebbd8c0eb425674f6fd6ddbb8118

Cryoris commented on crates/cext/src/dag.rs:
```diff
@@ -1243,6 +1244,153 @@ pub unsafe extern "C" fn qk_dag_get_instruction(
+///
+/// ``other`` may include a smaller or equal number of wires for each type.
+///
+/// @param dag A pointer to be composed on.
```
```suggestion
/// @param dag A pointer to the DAG to be composed on.
```

Cryoris commented on crates/cext/src/dag.rs:
```diff
@@ -1243,6 +1244,153 @@ pub unsafe extern "C" fn qk_dag_get_instruction(
+/// //        ┌─┴─┐└───┘└────────┘
+/// // rqr_2: ┤ X ├───────────────
+/// //        └───┘               
+/// qk_dag_compose(dag_right, dag_left, NULL, 0, NULL, 0)
```
```suggestion
/// qk_dag_compose(dag_right, dag_left, NULL, 0, NULL, 0);
```

Cryoris commented on crates/cext/src/dag.rs:
```diff
@@ -1243,6 +1244,153 @@ pub unsafe extern "C" fn qk_dag_get_instruction(
+///     onto.
+/// @param num_clbits The number of qubits in ``clbits``.
+///
+/// @return an exit code.
```
It would be nice to be slightly more explicit about the exit codes that could be returned 🙂

Cryoris commented on crates/cext/src/dag.rs:
```diff
@@ -1243,6 +1244,153 @@ pub unsafe extern "C" fn qk_dag_get_instruction(
+    };
+
+    let clbits: Option<Vec<ShareableClbit>> = if check_ptr(clbits).is_ok() {
+        let clbits = unsafe { std::slice::from_raw_parts(clbits, num_clbits as usize) };
```
This misses a SAFETY comment

Cryoris commented on crates/cext/src/dag.rs:
```diff
@@ -1243,6 +1244,153 @@ pub unsafe extern "C" fn qk_dag_get_instruction(
+    }
+
+    let qubits: Option<Vec<ShareableQubit>> = if check_ptr(qubits).is_ok() {
+        let qubits = unsafe { std::slice::from_raw_parts(qubits, num_qubits as usize) };
```
Safety comment 🙂

Cryoris commented on crates/cext/src/dag.rs:
```diff
@@ -1243,6 +1244,153 @@ pub unsafe extern "C" fn qk_dag_get_instruction(
+    num_clbits: u32,
+) -> ExitCode {
+    // SAFETY: Per documentation, the pointer is to valid data.
+
```
```suggestion
```

raynelfss committed:
Apply suggestions from code review

Co-authored-by: Julien Gacon <gaconju@gmail.com>

Cryoris commented on crates/cext/src/dag.rs:
```diff
@@ -1243,6 +1244,153 @@ pub unsafe extern "C" fn qk_dag_get_instruction(
+        let new_qubits: Result<Vec<ShareableQubit>, ExitCode> = qubits
+            .iter()
+            .map(|bit| -> Result<ShareableQubit, ExitCode> {
+                dag.qubits()
```
Nothing you have to change but you're using different flows for the qubits and clbits here, for doing the same thing, no? I think it would be nice to use the same here for easier readability / maintainability. I would have a slight preference for the clbit case below, which seems a bit easier to read.

raynelfss committed:
FIx: Remove `num_qubits` and `num_clbits`
- Fix error message for general failure in compose.

raynelfss committed:
Apply suggestion from @Cryoris

Co-authored-by: Julien Gacon <gaconju@gmail.com>

Cryoris committed:
Small docs & test fixes

- expand exit code return
- fix goto statement missing expected_dag
- improve consistency treating qubit and clbit

Cryoris reviewed (approved):
LGTM, thanks!
